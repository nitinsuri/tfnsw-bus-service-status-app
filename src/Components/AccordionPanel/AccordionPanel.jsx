import React, { useState } from 'react';
import Alerts from '../Alert/Alert';
import { deviationLabel, highlightSubString } from '../../utils';
import { appStaticLables } from '../../constants';
import Accordion from '@mui/material/Accordion';
import AccordionSummary from '@mui/material/AccordionSummary';
import AccordionDetails from '@mui/material/AccordionDetails';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import Chip from '@mui/material/Chip';
import Divider from '@mui/material/Divider';
import Stack from '@mui/material/Stack';

export default function AccordionPanel(props) {
  const { service, idx, expanded, handleAccordionChange } = props;

  const {
    sectionHeaders,
    textAreaPlaceholder,
    saveNotesButton,
    confirmSaveMessage,
  } = appStaticLables;

  const savedNote =
    localStorage.getItem(`note${idx}`)?.length > 0
      ? localStorage.getItem(`note${idx}`)
      : '';

  const [note, setNote] = useState(savedNote);
  const [message, setMessage] = useState(false);

  function saveNoteHandler() {
    localStorage.setItem(`note${idx}`, note);
    setMessage(true);
    setTimeout(() => {
      setMessage(false);
    }, 2500);
  }
  return (
    <>
      <Accordion
        expanded={expanded === `panel${idx}`}
        onChange={handleAccordionChange(`panel${idx}`)}
      >
        <AccordionSummary
          expandIcon={<ExpandMoreIcon />}
          aria-controls={`panel${idx}a-content`}
          id={`panel${idx}a-header`}
        >
          <strong>{service.organisation}</strong> - {service.date}
        </AccordionSummary>
        <AccordionDetails>
          <ul className="flat-list">
            <li className="list-header">
              <span>{sectionHeaders.busId}</span>
              <span>{sectionHeaders.variant}</span>
              <span>{sectionHeaders.deviation}</span>
            </li>
            {service.busData.map((serviceData) => {
              const { busId, routeVariant, deviationFromTimetable } =
                serviceData;
              const { label, status } = deviationLabel(deviationFromTimetable);
              console.log(label, status);
              const rvLabel = highlightSubString(routeVariant, 0, 3);
              return (
                <li key={busId}>
                  <span>{busId}</span>
                  <span dangerouslySetInnerHTML={{ __html: rvLabel }}></span>
                  <span>
                    <Chip
                      label={label}
                      color={status}
                      size="small"
                      variant="outlined"
                    />
                  </span>
                </li>
              );
            })}
          </ul>
          <Stack spacing={2}>
            {message && (
              <Alerts severity={'info'} message={confirmSaveMessage} />
            )}
            <textarea
              id={`panel${idx}a-header-notes`}
              placeholder={textAreaPlaceholder}
              onChange={(e) => setNote(e.target.value)}
              defaultValue={note}
            />
            <button onClick={saveNoteHandler}>{saveNotesButton}</button>
          </Stack>
        </AccordionDetails>
      </Accordion>
      <Divider />
    </>
  );
}
