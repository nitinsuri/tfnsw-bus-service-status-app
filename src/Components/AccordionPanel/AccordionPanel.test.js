import { render } from '@testing-library/react'
import AccordionPanel from './AccordionPanel'

test('loads and displays the first AccordionPanel', () => {
  const props = {
    service: {
      organisation: 'Westbus',
      date: '27/09/2023',
      busData: [
        {
          busId: '42612',
          routeVariant: '891 2 1',
          deviationFromTimetable: -77
        },
        {
          busId: '29016',
          routeVariant: '400 1 1',
          deviationFromTimetable: 0
        },
        {
          busId: '94811',
          routeVariant: '664 2 1',
          deviationFromTimetable: 164
        },
        {
          busId: '62788',
          routeVariant: 'UNKNOWN',
          deviationFromTimetable: null
        },
        {
          busId: '14221',
          routeVariant: '834 1 1',
          deviationFromTimetable: 423
        }
      ]
    },
    idx: 0,
    expanded: 'panel0',
    handleAccordionChange: jest.fn()
  }
  const accordionPanel = render(<AccordionPanel {...props} />)
  expect(accordionPanel).toMatchSnapshot()
})
