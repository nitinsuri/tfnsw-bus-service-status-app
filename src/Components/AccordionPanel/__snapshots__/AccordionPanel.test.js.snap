// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loads and displays the first AccordionPanel 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded Mui-expanded MuiAccordion-gutters css-1elwnq4-MuiPaper-root-MuiAccordion-root"
      >
        <div
          aria-controls="panel0a-content"
          aria-expanded="true"
          class="MuiButtonBase-root MuiAccordionSummary-root Mui-expanded MuiAccordionSummary-gutters css-sh22l5-MuiButtonBase-root-MuiAccordionSummary-root"
          id="panel0a-header"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiAccordionSummary-content Mui-expanded MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
          >
            <strong>
              Westbus
            </strong>
             - 
            27/09/2023
          </div>
          <div
            class="MuiAccordionSummary-expandIconWrapper Mui-expanded css-yw020d-MuiAccordionSummary-expandIconWrapper"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ExpandMoreIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
          </div>
        </div>
        <div
          class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-pwcg7p-MuiCollapse-root"
          style="min-height: 0px;"
        >
          <div
            class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
          >
            <div
              class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
            >
              <div
                aria-labelledby="panel0a-header"
                class="MuiAccordion-region"
                id="panel0a-content"
                role="region"
              >
                <div
                  class="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
                >
                  <ul
                    class="flat-list"
                  >
                    <li
                      class="list-header"
                    >
                      <span>
                        Bus Id
                      </span>
                      <span>
                        Route Variant
                      </span>
                      <span>
                        Status
                      </span>
                    </li>
                    <li>
                      <span>
                        42612
                      </span>
                      <span>
                        <span
                          class="highlight"
                        >
                          891
                        </span>
                         2 1
                      </span>
                      <span>
                        <div
                          class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorInfo MuiChip-outlinedInfo css-9m8jbz-MuiChip-root"
                        >
                          <span
                            class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
                          >
                            Late
                          </span>
                        </div>
                      </span>
                    </li>
                    <li>
                      <span>
                        29016
                      </span>
                      <span>
                        <span
                          class="highlight"
                        >
                          400
                        </span>
                         1 1
                      </span>
                      <span>
                        <div
                          class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorSuccess MuiChip-outlinedSuccess css-jdl1l1-MuiChip-root"
                        >
                          <span
                            class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
                          >
                            On-time
                          </span>
                        </div>
                      </span>
                    </li>
                    <li>
                      <span>
                        94811
                      </span>
                      <span>
                        <span
                          class="highlight"
                        >
                          664
                        </span>
                         2 1
                      </span>
                      <span>
                        <div
                          class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorError MuiChip-outlinedError css-47r1w-MuiChip-root"
                        >
                          <span
                            class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
                          >
                            Early
                          </span>
                        </div>
                      </span>
                    </li>
                    <li>
                      <span>
                        62788
                      </span>
                      <span>
                        UNKNOWN
                      </span>
                      <span>
                        <div
                          class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorDefault MuiChip-outlinedDefault css-1gyxgoa-MuiChip-root"
                        >
                          <span
                            class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
                          >
                            Unknown
                          </span>
                        </div>
                      </span>
                    </li>
                    <li>
                      <span>
                        14221
                      </span>
                      <span>
                        <span
                          class="highlight"
                        >
                          834
                        </span>
                         1 1
                      </span>
                      <span>
                        <div
                          class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorError MuiChip-outlinedError css-47r1w-MuiChip-root"
                        >
                          <span
                            class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
                          >
                            Early
                          </span>
                        </div>
                      </span>
                    </li>
                  </ul>
                  <div
                    class="MuiStack-root css-1sazv7p-MuiStack-root"
                  >
                    <textarea
                      id="panel0a-header-notes"
                      placeholder="Enter notes here. You can save this note for future by hitting the SAVE NOTES button"
                    />
                    <button
                      class="save-notes-button"
                    >
                      SAVE NOTES
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded Mui-expanded MuiAccordion-gutters css-1elwnq4-MuiPaper-root-MuiAccordion-root"
    >
      <div
        aria-controls="panel0a-content"
        aria-expanded="true"
        class="MuiButtonBase-root MuiAccordionSummary-root Mui-expanded MuiAccordionSummary-gutters css-sh22l5-MuiButtonBase-root-MuiAccordionSummary-root"
        id="panel0a-header"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiAccordionSummary-content Mui-expanded MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
        >
          <strong>
            Westbus
          </strong>
           - 
          27/09/2023
        </div>
        <div
          class="MuiAccordionSummary-expandIconWrapper Mui-expanded css-yw020d-MuiAccordionSummary-expandIconWrapper"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ExpandMoreIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </div>
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-pwcg7p-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              aria-labelledby="panel0a-header"
              class="MuiAccordion-region"
              id="panel0a-content"
              role="region"
            >
              <div
                class="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
              >
                <ul
                  class="flat-list"
                >
                  <li
                    class="list-header"
                  >
                    <span>
                      Bus Id
                    </span>
                    <span>
                      Route Variant
                    </span>
                    <span>
                      Status
                    </span>
                  </li>
                  <li>
                    <span>
                      42612
                    </span>
                    <span>
                      <span
                        class="highlight"
                      >
                        891
                      </span>
                       2 1
                    </span>
                    <span>
                      <div
                        class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorInfo MuiChip-outlinedInfo css-9m8jbz-MuiChip-root"
                      >
                        <span
                          class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
                        >
                          Late
                        </span>
                      </div>
                    </span>
                  </li>
                  <li>
                    <span>
                      29016
                    </span>
                    <span>
                      <span
                        class="highlight"
                      >
                        400
                      </span>
                       1 1
                    </span>
                    <span>
                      <div
                        class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorSuccess MuiChip-outlinedSuccess css-jdl1l1-MuiChip-root"
                      >
                        <span
                          class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
                        >
                          On-time
                        </span>
                      </div>
                    </span>
                  </li>
                  <li>
                    <span>
                      94811
                    </span>
                    <span>
                      <span
                        class="highlight"
                      >
                        664
                      </span>
                       2 1
                    </span>
                    <span>
                      <div
                        class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorError MuiChip-outlinedError css-47r1w-MuiChip-root"
                      >
                        <span
                          class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
                        >
                          Early
                        </span>
                      </div>
                    </span>
                  </li>
                  <li>
                    <span>
                      62788
                    </span>
                    <span>
                      UNKNOWN
                    </span>
                    <span>
                      <div
                        class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorDefault MuiChip-outlinedDefault css-1gyxgoa-MuiChip-root"
                      >
                        <span
                          class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
                        >
                          Unknown
                        </span>
                      </div>
                    </span>
                  </li>
                  <li>
                    <span>
                      14221
                    </span>
                    <span>
                      <span
                        class="highlight"
                      >
                        834
                      </span>
                       1 1
                    </span>
                    <span>
                      <div
                        class="MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorError MuiChip-outlinedError css-47r1w-MuiChip-root"
                      >
                        <span
                          class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
                        >
                          Early
                        </span>
                      </div>
                    </span>
                  </li>
                </ul>
                <div
                  class="MuiStack-root css-1sazv7p-MuiStack-root"
                >
                  <textarea
                    id="panel0a-header-notes"
                    placeholder="Enter notes here. You can save this note for future by hitting the SAVE NOTES button"
                  />
                  <button
                    class="save-notes-button"
                  >
                    SAVE NOTES
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
